src/com/palmergames/bukkit/towny/object/TownBlockTypeCache.java

```diff
@@ -0,0 +1,89 @@
+package com.palmergames.bukkit.towny.object;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class TownBlockTypeCache {
+
+	private Map<TownBlockType, Integer> typeCache = new ConcurrentHashMap<>();
+	private Map<TownBlockType, Integer> forSaleCache = new ConcurrentHashMap<>();
+	private Map<TownBlockType, Integer> residentOwnedCache = new ConcurrentHashMap<>();
+	
+	/*
+	 * Type Cache methods
+	 */
+
+	public int getNumTownBlocksOfType(TownBlockType type) {
+		return typeCache.containsKey(type) ? typeCache.get(type) : 0;
+	}
+
+	public void removeTownBlockOfType(TownBlock townBlock) {
+		typeCache.merge(townBlock.getType(), -1, Integer::sum);
+	}
+
+	public void addTownBlockOfType(TownBlock townBlock) {
+		typeCache.merge(townBlock.getType(), 1, Integer::sum);
+	}
+
+	public void removeTownBlockOfType(TownBlockType type) {
+		typeCache.merge(type, -1, Integer::sum);
+	}
+
+	public void addTownBlockOfType(TownBlockType type) {
+		typeCache.merge(type, 1, Integer::sum);
+	}
+
+	/*
+	 * ForSale by Type Cache methods
+	 */
+
+	public int getNumTownBlocksOfTypeForSale(TownBlockType type) {

```

Comment:
Instead of long method names, I think getNumTownBlocks(TownBlockType, CacheType) would be more readable. The cachetype enum would have 3 values: ALL, FOR_SALE, RESIDENT_OWNED